firebase-firestore/src/test/java/com/google/firebase/firestore/model/MutationSquashTest.java

```diff
@@ -0,0 +1,497 @@
+    patch.applyToLocalView(doc, now);
+
+    Mutation merge =
+        mergeMutation(
+            "collection/key",
+            map("arrays", FieldValue.arrayUnion(1, 2, 3)),
+            Arrays.asList(field("arrays")));
+    merge.applyToLocalView(doc, now);
+
+    Mutation squashed = emptyMutation("collection/key");
+    squashed = patch.squash(squashed, original, now);
+    squashed = merge.squash(squashed, original, now);
+    squashed.applyToLocalView(original, now);
+    assertEquals(doc, original);
+  }
+
+  @Test
+  public void testSquashArrayUnionThenRemove() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    MutableDocument original = doc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation union =
+        mergeMutation(
+            "collection/key", map("arrays", FieldValue.arrayUnion(1, 2, 3)), Arrays.asList());
+    Mutation remove =
+        mergeMutation(
+            "collection/key",
+            map("foo", "xxx", "arrays", FieldValue.arrayRemove(2)),
+            Arrays.asList(field("foo")));
+    union.applyToLocalView(doc, now);
+    remove.applyToLocalView(doc, now);
+
+    Mutation squashed = emptyMutation("collection/key");
+    squashed = union.squash(squashed, original, now);
+    squashed = remove.squash(squashed, original, now);
+    squashed.applyToLocalView(original, now);
+    assertEquals(doc, original);
+  }
+
+  @Test
+  public void testSquashSetThenIncrement() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    MutableDocument original = doc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation set = setMutation("collection/key", map("foo", 2));
+    Mutation update = patchMutation("collection/key", map("foo", FieldValue.increment(2)));
+    set.applyToLocalView(doc, now);
+    update.applyToLocalView(doc, now);
+
+    Mutation squashed = emptyMutation("collection/key");
+    squashed = set.squash(squashed, original, now);
+    squashed = update.squash(squashed, original, now);
+    squashed.applyToLocalView(original, now);
+    assertEquals(doc, original);
+  }
+
+  @Test
+  public void testSquashSetThenPatchOnDeletedDoc() {
+    MutableDocument doc = deletedDoc("collection/key", 0);
+    MutableDocument original = doc.clone();
+
+    Timestamp now = Timestamp.now();
+    Mutation set = setMutation("collection/key", map("bar", "bar-value"));
+    Mutation patch =
+        patchMutation(
+            "collection/key",
+            map("foo", "foo-patched-value", "bar.baz", FieldValue.serverTimestamp()));
+    set.applyToLocalView(doc, now);
+    patch.applyToLocalView(doc, now);
+
+    Mutation squashed = emptyMutation("collection/key");
+    squashed = set.squash(squashed, original, now);
+    squashed = patch.squash(squashed, original, now);
+    squashed.applyToLocalView(original, now);
+    assertEquals(doc, original);
+  }
+
+  // TODO(overlay): This test fails because deleting nested fields like bar.baz leaves an empty bar
```

Comment:
@schmidt-sebastian This currently fails, but I do not think this is an issue, we should discuss this.

Reply:
I think this is an issue. This changes the behavior - we should be left with an empty map. I don't (yet) know how to solve this though.